src/tap/src/tapcell.tcl

```diff
@@ -672,18 +683,19 @@ proc insert_at_top_bottom {db rows masters endcap_master prefix} {
       set x_start [expr $llx+$endcapwidth]
       set x_end [expr $urx-$endcapwidth]
 
-      insert_at_top_bottom_helper $block $cur_row $ori $x_start $x_end $lly $tap_nwintie_master $tap_nwin2_master $tap_nwin3_master $tap_nwouttie_master $tap_nwout2_master $tap_nwout3_master $prefix
+      insert_at_top_bottom_helper $block $cur_row 0 $ori $x_start $x_end $lly $tap_nwintie_master $tap_nwin2_master $tap_nwin3_master $tap_nwouttie_master $tap_nwout2_master $tap_nwout3_master $prefix
     }
   }
   set topbottom_cnt [expr $phy_idx - $start_phy_idx]
   utl::info TAP 6 "Top/bottom cells inserted: $topbottom_cnt"
   return $topbottom_cnt
 }
 
-proc insert_at_top_bottom_helper {block top_bottom ori x_start x_end lly tap_nwintie_master tap_nwin2_master tap_nwin3_master tap_nwouttie_master tap_nwout2_master tap_nwout3_master prefix} {
+proc insert_at_top_bottom_helper {block top_bottom is_macro ori x_start x_end lly tap_nwintie_master tap_nwin2_master tap_nwin3_master tap_nwouttie_master tap_nwout2_master tap_nwout3_master prefix} {
   if {$top_bottom == 1} {
     # top
-    if { $ori == "MX" } {
+    if {($ori == "R0" && $is_macro) || \
+        ($ori == "MX" && !$is_macro)} {
```

Comment:
Would you explain this further - why are these particular orientations special?  Why does is_macro matter?

Reply:
tapcell is looking at row orientation here. After checking with Calibre, we've noticed that the master depended on where we're placing these top/bottom cells: at the top/bottom of the core area, or at the top/bottom of macros. With this update, Calibre reports 0 DRCs regarding tapcells.

Reply:
How does is_macro relate to row orientation?

Reply:
I want to make sure this is a general fix and not just for a specific issue.

Reply:
The cell we choose for placement at the top/bottom of rows and at the top/bottom of the core depends on the row orientation. For the core, we choose _IN_ cells when the orientation is MX, but for macros, we do it when the orientation is R0. In summary, the behavior of the top/bottom insertion depends if it is near macros or at the core.
I'm available to talk if it is not clear.